File: benchmarks/beginner/B3-array-sum/python/main.py
import time
import math


def main():
    N = 10_000_000
    RUNS = 10
    EXPECTED = 50000005000000
    timings = []

    print("=== BENCHMARK: B3 - Array Sum ===")
    print("Language: Python")
    print(f"Runtime:  Python {__import__('sys').version.split()[0]}")
    print()

    for i in range(1, RUNS + 1):
        start = time.perf_counter()
        arr = list(range(1, N + 1))
        result = sum(arr)
        elapsed = (time.perf_counter() - start) * 1000  # ms
        timings.append(elapsed)
        print(f"Run {i:2d}: {elapsed:.2f} ms")

    print()
    min_t = min(timings)
    avg_t = sum(timings) / len(timings)
    max_t = max(timings)
    variance = sum((t - avg_t) ** 2 for t in timings) / len(timings)
    stddev = math.sqrt(variance)

    print(f"Min:    {min_t:.2f} ms")
    print(f"Avg:    {avg_t:.2f} ms")
    print(f"Max:    {max_t:.2f} ms")
    print(f"StdDev: {stddev:.2f} ms")
    print()
    status = "✓" if result == EXPECTED else "✗"
    print(f"Verification: {result} {status}")


if __name__ == "__main__":
    main()

File: benchmarks/beginner/B3-array-sum/go/main.go
package main

import (
	"fmt"
	"math"
	"time"
)

func main() {
	const N = 10_000_000
	const RUNS = 10
	const EXPECTED int64 = 50000005000000

	timings := make([]float64, 0, RUNS)

	fmt.Println("=== BENCHMARK: B3 - Array Sum ===")
	fmt.Println("Language: Go")
	fmt.Printf("Runtime:  Go\n")
	fmt.Println()

	var result int64
	for i := 1; i <= RUNS; i++ {
		start := time.Now()
		arr := make([]int64, N)
		for j := 0; j < N; j++ {
			arr[j] = int64(j + 1)
		}
		var s int64
		for _, v := range arr {
			s += v
		}
		result = s
		elapsed := float64(time.Since(start).Nanoseconds()) / 1e6 // ms
		timings = append(timings, elapsed)
		fmt.Printf("Run %2d: %.2f ms\n", i, elapsed)
	}

	fmt.Println()
	minT := timings[0]
	maxT := timings[0]
	sum := 0.0
	for _, t := range timings {
		sum += t
		if t < minT {
			minT = t
		}
		if t > maxT {
			maxT = t
		}
	}
	avg := sum / float64(len(timings))
	variance := 0.0
	for _, t := range timings {
		variance += (t - avg) * (t - avg)
	}
	variance /= float64(len(timings))
	stddev := math.Sqrt(variance)

	fmt.Printf("Min:    %.2f ms\n", minT)
	fmt.Printf("Avg:    %.2f ms\n", avg)
	fmt.Printf("Max:    %.2f ms\n", maxT)
	fmt.Printf("StdDev: %.2f ms\n", stddev)
	fmt.Println()
	status := "✗"
	if result == EXPECTED {
		status = "✓"
	}
	fmt.Printf("Verification: %d %s\n", result, status)
}
